_posts/2016-08-31-refactoring-rust-primitive-obsession.markdown

```diff
@@ -0,0 +1,205 @@
+
+We're not going to change the test, because refactoring should not change the behavior of existing code. Our function will still accept a string, but we can convert it to domain-specific Struct inside the function.
+
+What should we call this Struct? Naming is hard, as always. Since it contains a bunch of brackets let's just call it `Brackets` for now. Maybe a better name will occur to us later. `Brackets` will start off containing our input string.
+
+```rust
+struct Brackets {
+  input: String,
+};
+```
+
+How do we create our `Brackets` inside `are_balanced`?. We could do this:
+
+```rust
+pub fn are_balanced(input: String) -> bool {
+  let brackets = Brackets { input: input };
+  //...
+```
+
+And that's...OK. But it's not idiomatic. We're trying to convert a String to a Brackets and Rust has a standard way to convert types -- the [`From` trait](https://doc.rust-lang.org/std/convert/trait.From.html). You may have seen this trait before in the commonly used `String::from(....)` function. But you can implement `From` for any type!
+
+```rust
+struct Brackets {
+  input: String,
+};
+
+impl From<String> for Brackets {
+  fn from(input: String) -> Self {
+    Brackets { input: input }
+  }
+}
+```
+
+Wait! That's pretty much exactly the same thing we did in the last example. What's the benefit of implementing `From`? First, we're both following a Rust idiom and establishing a example for our code. And this example will guide us if (read: when) we want to create a Brackets from something else? A vector? A &str? Time will tell, but we've already established the protocol to follow whatever happens.
+
+All of this is good...but it doesn't help us. Our code still relies on the behavior of a string primitive:
+
+```rust
+pub fn are_balanced(input: String) -> bool {
+    //..
+    for current_bracket in input.chars() {
+    //..
+    }
+}
+```
+
+Our `Brackets` doesn't implement `chars()`. We can make it do so:
+
+```rust
+use std::str::Chars;
+impl Brackets {
+  fn brackets(&self) -> Chars {
+    self.input.chars()
+  }
+}
+```
+
+That feels unsound to me, though I can't put a finger on why. I suspect this shortcut will come back and bite us later. Let's be a little more thorough and idiomatic.
+
+[`chars` returns an iterator](https://doc.rust-lang.org/std/primitive.str.html#method.chars) and -- just like the `From` trait and type conversion -- Rust has a standard way of adding iterators to your Structs, [`IntoIterator`](https://doc.rust-lang.org/core/iter/trait.IntoIterator.html)
+
+```rust
+impl IntoIterator for Brackets {
+  type Item = char;
+  type IntoIter = ::std::vec::IntoIter<char>;
+
+  fn into_iter(self) -> Self::IntoIter {
+    self.input.chars().collect::<Vec<char>>().into_iter()
```

Comment:
Thanks! I tried that but could not get the lifetime annotations correct and decided it in favor of the simpler-looking (if less efficient) vector approach.